streaming/stream_transfer_task.cc

```diff
@@ -105,6 +105,23 @@ struct send_info {
         , prs(to_partition_ranges(ranges))
         , reader(cf.make_streaming_reader(cf.schema(), prs)) {
     }
+    future<bool> has_relevant_range_on_this_shard() {
+        return do_with(false, [this] (bool& found_relevant_range) {
+            return do_for_each(ranges, [this, &found_relevant_range] (dht::token_range range) {
+                if (found_relevant_range) {
```

Comment:
Nitpick: you can put the logic below in if (!found_relevant_range) and drop both returns. Two lines less.
Although it is a pity we need to iterate over all ranges even if a first one belongs to the shard especially as you claim
there can be a lot of them.

Reply:
If we add "break" support to do_for_each, we can avoid the further loops.

Since this code is on a slow path, I do not worry about the extra loop a lot.

Reply:
We can add do_for_each version that takes a function that returns  future<stop_iteration>